### PR TITLE
params extraction using parse tree, regex as fallback

### LINKED ISSUE #322: Parameter Parser Duplicates for Oracle Queries
image

observed in the MTRAC query  SQL_SMS_R

...
where
    bdcuc.ROWID = bci_rowid 
    and
    ROWNUM <= 5 * :SQLT_INT_ROWLIMIT 
    and
    bci_submitdate >= :SQLT_DAT_START_DATE 
    and
    bci_submitdate <= :SQLT_DAT_END_DATE 
    and
      (
        bc_trycount in(0,1,2,3,5,9) 
        or
        bc_status in(4,5,16) 
      )
    and
      bc_msisdn_a like :SQLT_STR_SENDER 
    and
    bci_msisdn = :SQLT_STR_RECEIVER 
    and
      (
        :SQLT_STR_SHORTID is null 
        or
          BC_CONSOL like :SQLT_STR_SHORTID 
      )
    and
      (
        :SQLT_STR_GENERATED is null 
        or
            NVL(BC_GENERATED,0) 
          like
          :SQLT_STR_GENERATED
      )

### PR DESCRIPTION
**Pull request by c-bik**
Fixes #322
@acautin Need your help in testing this as well!

### CODE REVIEW

File: src/gen_adapter.erl [hunk 1]
```diff
@@ -54,9 +54,27 @@ add_cmds_views(Sess, UserId, A, Replace, [{N,C,Con,#viewstate{}=V}|Rest]) ->
...
-    case re:run(Sql, RegEx, [global,{capture, [0,1,2], binary}]) of
+    RegEx = "[^a-zA-Z0-9(]*:(" ++ string:join([binary_to_list(T) || T <- Types], "|")
+            ++ ")((_IN_|_OUT_|_INOUT_){0,1})[^ ,\)\n\r;]+",
+    case
+        try
+            {ok, PTree} = sqlparse:parsetree(Sql),
+            Params = sqlparse:foldtd(
+                       fun({param, P}, A) ->
+                               case re:run(
+                                      P, RegEx,
```
Comment (acautin) [inline src/gen_adapter.erl]: @c-bik why do we need the regexp at all, can we not trust the parsetree to give them separated ?

File: src/gen_adapter.erl [hunk 2]
```diff
@@ -54,9 +54,27 @@ add_cmds_views(Sess, UserId, A, Replace, [{N,C,Con,#viewstate{}=V}|Rest]) ->
...
+                                   _ -> A
+                               end;
+                          (_, A) -> A
+                       end, [], PTree),
+            {match, Params}
+        catch 
+            C:R ->
+                ?Warn("~p~n~p", [{C,R}, erlang:get_stacktrace()]),
+                re:run(Sql, RegEx, [global,{capture, [0,1,2], binary}])
+        end of
```
Comment (acautin) [inline src/gen_adapter.erl]: @c-bik this case inside a case looks really complicated...

File: src/gen_adapter.erl (hunk 1 above)
Comment (c-bik) [inline src/gen_adapter.erl]:
```erlang
> sqlparse:parsetree("select :SQLT_INT_IN_ABC from def").
{ok,[{{select,[{fields,[{param,<<":SQLT_INT_IN_ABC">>}]},
               {from,[<<"def">>]},
               {where,{}},
               {'hierarchical query',{}},
               {'group by',[]},
               {having,{}},
               {'order by',[]}]},
      {extra,<<>>}}]}
```
As you see, parser is not capable of symantic analysis of parameters string structure. `TYPE_DIR_NAME` format of SQL parametrs is a DDErl concept, and hence is the regexp to split `<<":SQLT_INT_IN_ABC">>` into `[<<":SQLT_INT_IN_ABC">>,<<":SQLT_INT">>,<<"_IN_">>]` etc.

File: src/gen_adapter.erl (hunk 2 above)
Comment (c-bik) [inline src/gen_adapter.erl]: Will try to simplify

Comment (c-bik) [inline src/gen_adapter.erl]: How about https://github.com/K2InformaticsGmbH/dderl/pull/325/commits/74ab3c1f8e8070d5dd96f60b33e1c7c76122a5ec?

**Pull request merged**